contributingGuides/TS_STYLE.md

```diff
@@ -0,0 +1,335 @@
+    // ... some complex logic here ...
+    return foo;
+  }
+  ```
+
+<a name="conventions-jsdoc"></a><a name="1.10"></a>
+
+- [1.10](#conventions-jsdoc) **JSDoc**: Omit comments that are redundant with TypeScript. Do not declare types in `@param` or `@return` blocks. Do not write `@implements`, `@enum`, `@private`, `@override`
+
+  ```ts
+  // bad
+  /**
+   * @param {number} age
+   * @returns {boolean} Whether the person is a legal drinking age or nots
+   */
+  function canDrink(age: number): boolean {
+    return age >= 21;
+  }
+
+  // good
+  /**
+   * @param age
+   * @returns Whether the person is a legal drinking age or nots
+   */
+  ```
+
+<a name="convension-proptypes-and-defaultprops"></a><a name="1.11"></a>
+
+- [1.11](#convension-proptypes-and-defaultprops) **`propTypes` and `defaultProps`**: Do not use them. Use object destructing to assign default values if necessary.
+
+  > Refer to [the propTypes Migration Table](./PROPTYPES_CONVERSION_TABLE.md) on how to type props based on existing `propTypes`.
+
+  ```tsx
+  type GreetingProps = {
+    greeting: string;
+    name: string;
+  };
+
+  function Greeting({ greeting = "hello", name = "world" }: ComponentProps) {
+    <Text>{`${greeting}, ${name}`}</Text>;
+  }
+  ```
+
+<a name="convension-utility-types"></a><a name="1.12"></a>
+
+- [1.12](#convension-utility-types) **Utility Types**: Use types from [TypeScript utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html) and [`type-fest`](https://github.com/sindresorhus/type-fest) when possible.
+
+  ```ts
+  type Foo = {
+    bar: string;
+  };
+
+  // good
+  type ReadOnlyFoo = Readonly<Foo>;
+
+  // bad
+  type ReadOnlyFoo = {
+    readonly [Property in keyof Foo]: Foo[Property];
+  };
+  ```
+
+<a name="convension-object-type"></a><a name="1.13"></a>
+
+- [1.13](#convension-object-type) **`object`**: Don't use `object` type.
+
+> Why? `object` refers to "any non-primitive type," not "any object". Typing "any non-primitive value" is not commonly needed.
+
+```ts
+// bad
+const foo: object = [1, 2, 3]; // TypeScript does not error
+```
+
+If you know that the type of data is an object but don't know what properties or values it has beforehand, use `Record<string, unknown>`.
+
+<a name="convension-export-prop-types"></a><a name="1.14"></a>
+
+- [1.14](#convension-export-prop-types) **Prop Types**: Define and export prop types for components. Use exported prop types instead of grabbing the prop type from a component.
+
+> Why? Exporting prop types aids reusability.
+
+```tsx
+// MyComponent.tsx
+export type MyComponentProps = {
+  foo: string;
+};
+
+export default function MyComponent({ foo }: MyComponentProps) {
+  return <Text>{foo}</Text>;
+}
+
+// bad
+import { ComponentProps } from "React";
+import MyComponent from "./MyComponent";
+type MyComponentProps = ComponentProps<typeof MyComponent>;
+
+// good
+import MyComponent, { MyComponentProps } from "./MyComponent";
+```
```

Comment:
@hayata-suenaga Are we going to always export the component's props, even when isn't necessary? cc @blazejkustra

Reply:
When props are only used inside of the file where it is declared I wouldn't export it.

Reply:
From previous guidelines PR:

> Types specific to a single file should remain inside that file to keep their scope limited and their usage clear.

Reply:
> When props are only used inside the file where it is declared, I wouldn't export it.

I was thinking it might be a good idea to always export a prop type if its component is also exported. We don't know if other components might need the prop type or not

Reply:
I don't have a strong opinion here but, it can lead to a sort of "premature optimization". We should not export props just in case they might be needed in the future (YAGNI principle). 

Instead, if such need arises where props are required in another component, we can start exporting them at that time. @hayata-suenaga

Reply:
I would prefer to not export every prop type as they are going to "pollute" the editor's autocomplete mechanisms unnecessarily, we should export only things that are going to be used across the project.

Reply:
okay that makes sense removing this part. should we allow the usage of `type MyComponentProps = ComponentProps<typeof MyComponent>;` or if the prop type is needed, should they be exported?

Reply:
@hayata-suenaga @blazejkustra I think it's better to export the prop type itself when needed instead of using `ComponentProps` 👍

Reply:
okay then let's include instructions to export and import prop types if you find yourself trying to use `ComponentProps`

what do you think @blazejkustra